hail/src/main/scala/is/hail/annotations/StagedRegionValueBuilder.scala

```diff
@@ -9,85 +9,28 @@ import is.hail.expr.types.virtual.{TBoolean, TFloat32, TFloat64, TInt32, TInt64,
 import is.hail.utils._
 
 object StagedRegionValueBuilder {
-  def fixupStruct(fb: EmitFunctionBuilder[_], region: Code[Region], typ: PBaseStruct, value: Code[Long]): Code[Unit] = {
-    coerce[Unit](Code(typ.fields.map { f =>
-      if (f.typ.isPrimitive)
-        Code._empty
-      else {
-        val fix = f.typ.fundamentalType match {
-          case t@(_: PBinary | _: PArray) =>
-            val off = fb.newField[Long]
-            Code(
-              off := typ.fieldOffset(value, f.index),
-              Region.storeAddress(off, deepCopyFromOffset(fb, region, t, coerce[Long](Region.loadIRIntermediate(t)(off))))
-            )
-          case t: PBaseStruct =>
-            val off = fb.newField[Long]
-            Code(off := typ.fieldOffset(value, f.index),
-              fixupStruct(fb, region, t, off))
-        }
-        typ.isFieldDefined(value, f.index).mux(fix, Code._empty)
-      }
-    }: _*))
-  }
-
-  def fixupArray(fb: EmitFunctionBuilder[_], region: Code[Region], typ: PArray, value: Code[Long]): Code[Unit] = {
-    if (typ.elementType.isPrimitive)
-      return Code._empty
-
-    val i = fb.newField[Int]
-    val len = fb.newField[Int]
-
-    val perElt = typ.elementType.fundamentalType match {
-      case t@(_: PBinary | _: PArray) =>
-        val off = fb.newField[Long]
-        Code(
-          off := typ.elementOffset(value, len, i),
-          Region.storeAddress(off, deepCopyFromOffset(fb, region, t, coerce[Long](Region.loadIRIntermediate(t)(off)))))
-      case t: PBaseStruct =>
-        val off = fb.newField[Long]
-        Code(off := typ.elementOffset(value, len, i),
-          fixupStruct(fb, region, t, off))
-    }
-    Code(
-      i := 0,
-      len := typ.loadLength(value),
-      Code.whileLoop(i < len,
-        typ.isElementDefined(value, i).mux(perElt, Code._empty),
-        i := i + 1))
-  }
-
   def deepCopy(fb: EmitFunctionBuilder[_], region: Code[Region], typ: PType, value: Code[_], dest: Code[Long]): Code[Unit] = {
-    typ.fundamentalType match {
-      case t if t.isPrimitive => Region.storePrimitive(t, dest)(value)
-      case t@(_: PBinary | _: PArray) =>
-        Region.storeAddress(dest, deepCopyFromOffset(fb, region, t, coerce[Long](value)))
-      case t: PBaseStruct =>
-        Code(Region.copyFrom(coerce[Long](value), dest, t.byteSize),
-          fixupStruct(fb, region, t, dest))
-      case t => fatal(s"unknown type $t")
+    val t = typ.fundamentalType
+    val valueTI = ir.typeToTypeInfo(t)
+    val mb = fb.getOrDefineMethod("deepCopy", ("deepCopy", typ),
+      Array[TypeInfo[_]](classInfo[Region], valueTI, LongInfo), UnitInfo) { mb =>
+      val r = mb.getArg[Region](1)
+      val value = mb.getArg(2)(valueTI)
+      val dest = mb.getArg[Long](3)
+      mb.emit(t.constructAtAddressFromValue(mb, dest, r, t, value, true))
```

Comment:
Could you help me understand why we need to emit and invoke here? It doesn’t appear we were emitting previously, just constructing bytecode by performing the equivalent of deep copyFromType

Reply:
slight preference for deepCopy = true, may be easier to read

Reply:
This will generate smaller methods (potentially improved performance) and makes our decision to use locals instead of fields in the ptype copy methods totally safe when called from this entry point.

Another benefit is that we'll generate less code total -- we memoize the method for a given ptype/ptype combination, and only generate it once.

The emit/invoke is just how method builders work. You emit code into them to define the method, and `mb.invoke(args)` returns a `Code[_]` that emits bytecode for *call this method with these args*.

Reply:
That seems like a great change. I though that smaller method size could have been the reason, wanted to confirm. Thanks, for that and the explanation!